File: api/generate.js
const JSZip = require('jszip');
const fs = require('fs');
const path = require('path');

// Text replacements: template text → field key from request body
const REPLACEMENTS = [
  { from: 'Creación de Contenido (Video & Foto)+ Brand assesement ', key: 'nombre_servicio' },
  { from: 'Creación de Contenido (Video & Foto)+ Brand assesement', key: 'nombre_servicio' },
  { from: 'Creación de contenido audiovisual en formato vertical, optimizado para redes sociales, alineado con la estrategia de la marca y pensado para facilitar su ejecución y publicación.', key: 'descripcion' },
  { from: 'Acompañamiento estratégico para ayudar a una marca a identificar su posición real en el mercado, clarificar su voz y diseñar una hoja de ruta de contenido accionable. A través de un proceso estructurado de reuniones y feedback, definimos dónde está la marca hoy, dónde debería estar, y cómo llegar ahí', key: 'incluye' },
  { from: 'Plazo: 6 semanas', key: 'plazo', prefix: 'Plazo: ' },
  { from: '6 semanas', key: 'plazo' },
  { from: "Brand Discovery Document: análisis de competidores, oportunidad en el mercado digital, tono de voz, do's &amp; don'ts", key: 'entregable_1' },
  { from: 'Content Guide (DIY): guía de contenido que el propio equipo de la marca puede producir', key: 'entregable_2' },
  { from: 'Filmación y edición de 6 videos (Reels) con contenido previamente definido, ', key: 'entregable_3' },
  { from: 'Filmación y edición de 6 videos (Reels) con contenido previamente definido,', key: 'entregable_3' },
  { from: '60 fotografías de producto y de situación, pensadas para feed, stories y otros usos digitales.', key: 'entregable_4' },
  { from: 'Dirección de arte, asegurando coherencia visual, estética de marca y calidad en todo el contenido.', key: 'entregable_5' },
  { from: 'CON ESTE PACK SE PREVEE CUBRIR EL CONTENIDO PARA UN MÍNIMO DE 3 MESES.', key: 'nota_aclaracion' },
  { from: '2.500€', key: 'precio' },
  { from: '2.500&#x20AC;', key: 'precio' },
  { from: 'Brand discovery', key: 'timeline_1' },
  { from: 'Producción contenido', key: 'timeline_2' },
  { from: 'Entrega contenidos', key: 'timeline_3' },
];

function escapeXml(str) {
  return str
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;');
}

function replaceInXml(xml, from, to) {
  // Direct string replacement (XML-safe)
  let result = xml;
  // Try literal replacement first
  if (result.includes(from)) {
    result = result.split(from).join(to);
  }
  // Try with XML-escaped version
  const escapedFrom = escapeXml(from);
  if (escapedFrom !== from && result.includes(escapedFrom)) {
    result = result.split(escapedFrom).join(escapeXml(to));
  }
  return result;
}

module.exports = async (req, res) => {
  // CORS
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }

  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  try {
    const data = req.body;

    // Load template
    const templatePath = path.join(__dirname, 'template.pptx');
    const templateBuffer = fs.readFileSync(templatePath);

    // Open as ZIP
    const zip = await JSZip.loadAsync(templateBuffer);

    // Process all slide XML files
    const slideFiles = Object.keys(zip.files).filter(name =>
      name.startsWith('ppt/slides/slide') && name.endsWith('.xml')
    );

    for (const slideFile of slideFiles) {
      let xml = await zip.files[slideFile].async('string');

      // Apply all replacements
      for (const { from, key, prefix } of REPLACEMENTS) {
        const val = data[key];
        if (!val) continue;
        const to = prefix ? prefix + val : val;
        xml = replaceInXml(xml, from, to);
      }

      zip.file(slideFile, xml);
    }

    // Generate output
    const outputBuffer = await zip.generateAsync({
      type: 'nodebuffer',
      compression: 'DEFLATE',
      compressionOptions: { level: 6 }
    });

    const filename = `Bright_Propuesta_${(data.nombre_servicio || 'Servicio').replace(/[^a-zA-Z0-9]/g, '_').slice(0, 30)}.pptx`;

    res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.presentationml.presentation');
    res.setHeader('Content-Disposition', `attachment; filename="${filename}"`);
    res.setHeader('Content-Length', outputBuffer.length);

    return res.status(200).send(outputBuffer);

  } catch (err) {
    console.error('Generate error:', err);
    return res.status(500).json({ error: err.message });
  }
};

File: api/generate.py
from http.server import BaseHTTPRequestHandler
import json
import os
import io
import base64
from pptx import Presentation
from pptx.util import Pt
import copy

TEMPLATE_PATH = os.path.join(os.path.dirname(__file__), "template.pptx")

REPLACEMENTS = {
    "{{nombre_servicio}}": "nombre_servicio",
    "{{descripcion}}":     "descripcion",
    "{{plazo}}":           "plazo",
    "{{incluye}}":         "incluye",
    "{{entregable_1}}":    "entregable_1",
    "{{entregable_2}}":    "entregable_2",
    "{{entregable_3}}":    "entregable_3",
    "{{entregable_4}}":    "entregable_4",
    "{{entregable_5}}":    "entregable_5",
    "{{nota_aclaracion}}": "nota_aclaracion",
    "{{precio}}":          "precio",
    "{{timeline_1}}":      "timeline_1",
    "{{timeline_2}}":      "timeline_2",
    "{{timeline_3}}":      "timeline_3",
}

# These are the actual texts in Gabi's template that map to each field
TEMPLATE_TEXTS = {
    "nombre_servicio": "Creación de Contenido (Video & Foto)+ Brand assesement",
    "descripcion":     "Creación de contenido audiovisual en formato vertical, optimizado para redes sociales, alineado con la estrategia de la marca y pensado para facilitar su ejecución y publicación.",
    "plazo":           "6 semanas",
    "incluye":         "Acompañamiento estratégico para ayudar a una marca a identificar su posición real en el mercado, clarificar su voz y diseñar una hoja de ruta de contenido accionable. A través de un proceso estructurado de reuniones y feedback, definimos dónde está la marca hoy, dónde debería estar, y cómo llegar ahí",
    "entregable_1":    "Brand Discovery Document: análisis de competidores, oportunidad en el mercado digital, tono de voz, do's & don'ts",
    "entregable_2":    "Content Guide (DIY): guía de contenido que el propio equipo de la marca puede producir",
    "entregable_3":    "Filmación y edición de 6 videos (Reels) con contenido previamente definido,",
    "entregable_4":    "60 fotografías de producto y de situación, pensadas para feed, stories y otros usos digitales.",
    "entregable_5":    "Dirección de arte, asegurando coherencia visual, estética de marca y calidad en todo el contenido.",
    "nota_aclaracion": "CON ESTE PACK SE PREVEE CUBRIR EL CONTENIDO PARA UN MÍNIMO DE 3 MESES.",
    "precio":          "2.500€",
    "timeline_1":      "Brand discovery",
    "timeline_2":      "Producción contenido",
    "timeline_3":      "Entrega contenidos",
}

def replace_in_run(run, old, new):
    if old in run.text:
        run.text = run.text.replace(old, new)
        return True
    return False

def replace_in_shape(shape, data):
    if not shape.has_text_frame:
        return
    for para in shape.text_frame.paragraphs:
        for run in para.runs:
            for field, key in TEMPLATE_TEXTS.items():
                val = data.get(key, "")
                if val and field in run.text:
                    run.text = run.text.replace(field, val)

def replace_text_in_presentation(prs, data):
    """Replace template texts with user data in all slides."""
    for slide in prs.slides:
        for shape in slide.shapes:
            if shape.has_text_frame:
                for para in shape.text_frame.paragraphs:
                    # Work at paragraph level to handle split runs
                    full_text = "".join(r.text for r in para.runs)
                    for template_text, field_key in TEMPLATE_TEXTS.items():
                        new_val = data.get(field_key, "")
                        if new_val and template_text in full_text:
                            # Replace in first run that contains part of the text
                            for run in para.runs:
                                if template_text in run.text:
                                    run.text = run.text.replace(template_text, new_val)
                                    break
                            # Handle split across runs - reconstruct
                            full_text = full_text.replace(template_text, new_val)

def generate_pptx(data):
    prs = Presentation(TEMPLATE_PATH)
    replace_text_in_presentation(prs, data)
    buf = io.BytesIO()
    prs.save(buf)
    buf.seek(0)
    return buf.read()

class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.end_headers()

    def do_POST(self):
        try:
            length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(length)
            data = json.loads(body)

            pptx_bytes = generate_pptx(data)
            b64 = base64.b64encode(pptx_bytes).decode("utf-8")

            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.send_header("Access-Control-Allow-Origin", "*")
            self.end_headers()
            self.wfile.write(json.dumps({"file": b64, "filename": f"Bright_Propuesta_{data.get('nombre_servicio','').replace(' ','_')[:30]}.pptx"}).encode())

        except Exception as e:
            self.send_response(500)
            self.send_header("Content-Type", "application/json")
            self.send_header("Access-Control-Allow-Origin", "*")
            self.end_headers()
            self.wfile.write(json.dumps({"error": str(e)}).encode())
